dev/io.openliberty.concurrent.internal/src/io/openliberty/concurrent/internal/processor/ConcurrencyResourceFactoryBuilder.java

```diff
@@ -13,27 +13,73 @@
 package io.openliberty.concurrent.internal.processor;
 
 import java.lang.annotation.Annotation;
+import java.util.Arrays;
 
+import com.ibm.websphere.ras.Tr;
+import com.ibm.websphere.ras.TraceComponent;
 import com.ibm.ws.resource.ResourceFactoryBuilder;
+import com.ibm.ws.runtime.metadata.ApplicationMetaData;
+import com.ibm.ws.runtime.metadata.ComponentMetaData;
+import com.ibm.ws.runtime.metadata.MetaData;
+import com.ibm.ws.runtime.metadata.ModuleMetaData;
 
 /**
- * Super interface for all ResourceFcatoryBuilders that are provided by the
+ * Superclass for all ResourceFactoryBuilders that are provided by the
  * Concurrency component.
  */
-public interface ConcurrencyResourceFactoryBuilder extends ResourceFactoryBuilder {
+abstract class ConcurrencyResourceFactoryBuilder //
+                implements ResourceFactoryBuilder {
+    private static final TraceComponent tc = //
+                    Tr.register(ConcurrencyResourceFactoryBuilder.class);
+
+    /**
+     * Create an exception for the error condition where a java:global name is
+     * configured on an application-defined resource that also configures
+     * qualifiers.
+     *
+     * @param jndiName   JNDI name of the resource definition.
+     * @param qualifiers qualifier names.
+     * @param appName    application name.
+     * @param metadata   metadata of the application artifact.
+     * @return UnsupportedOperationException.
+     */
+    UnsupportedOperationException excJavaGlobalWithQualifiers(String jndiName,
+                                                              String[] qualifiers,
+                                                              String appName,
+                                                              MetaData metadata) {
+        String artifactName;
+        if (metadata instanceof ComponentMetaData)
+            artifactName = ((ComponentMetaData) metadata).getJ2EEName().toString();
+        else if (metadata instanceof ModuleMetaData)
+            artifactName = ((ModuleMetaData) metadata).getJ2EEName().toString();
+        else if (metadata instanceof ApplicationMetaData)
+            artifactName = ((ApplicationMetaData) metadata).getJ2EEName().toString();
```

Comment:
If you want to simplify this, based on the JavaDoc for `MetaData.getName()` this should be the same (I haven't tested it though):

>Gets the name associated with this meta data. For application, module and component metadata, this should be the same as the corresponding name in [getJ2EEName](eclipse-javadoc:%E2%98%82=com.ibm.ws.container.service.compat/src%3Ccom.ibm.ws.runtime.metadata%7BMetaData.java%E2%98%83MetaData~getName%E2%98%82%E2%98%82getJ2EEName).

```suggestion
        if (metadata instanceof ComponentMetaData || metadata instanceof ModuleMetaData || metadata instanceof ApplicationMetaData)
            artifactName = ((ComponentMetaData) metadata).getName();

```

Reply:
I did try with `getName()` before writing it this way and I remember it not returning all of the information, although I can't remember well enough to say specifically, so will need to check on it again.

Reply:
Confirmed that `getName()` sometimes returns `null` and sometimes returns only the module without the app name.  So we won't make this change.